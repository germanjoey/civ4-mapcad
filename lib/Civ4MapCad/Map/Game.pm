package Civ4MapCad::Map::Game;

use strict;
use warnings;

our @fields = qw(Era Speed Calendar GameTurn MaxTurns MaxCityElimination NumAdvancedStartPoints TargetScore StartYear Description ModPath);
our %field_names;
@field_names{@fields} = (1) x @fields;

use Civ4MapCad::Util qw(write_block_data);

sub new {
    my $proto = shift;
    my $class = ref $proto || $proto;
    
    my $obj = bless {}, $class;
    $obj->{'Victories'} = [];
    $obj->{'Options'} = [];
    $obj->{'MPOptions'} = [];
    $obj->{'ForceControls'} = [];
    
    return $obj;
}

sub new_default {
    my $proto = shift;
    my $class = ref $proto || $proto;
    my $obj = bless {}, $class;
    
    $obj->default();
    return $obj;
}

sub default {
    my ($self) = @_;
    $self->clear();
    
    $self->set('Era', 'ERA_ANCIENT');
    $self->set('Speed', 'GAMESPEED_NORMAL');
    $self->set('Calendar', 'CALENDAR_DEFAULT');
    
    $self->add_option('GAMEOPTION_LEAD_ANY_CIV');
    $self->add_option('GAMEOPTION_NO_VASSAL_STATES');
    $self->add_option('GAMEOPTION_NO_GOODY_HUTS');
    $self->add_option('GAMEOPTION_NO_EVENTS');
    
    $self->add_victory('VICTORY_TIME');
    $self->add_victory('VICTORY_CONQUEST');
    $self->add_victory('VICTORY_DOMINATION');
    $self->add_victory('VICTORY_CULTURAL');
    $self->add_victory('VICTORY_SPACE_RACE');
    $self->add_victory('VICTORY_DIPLOMATIC');
    
    $self->set('GameTurn', 0);
    $self->set('MaxTurns', 500);
    $self->set('MaxCityElimination', 0);
    $self->set('NumAdvancedStartPoints', 600);
    $self->set('TargetScore', 0);
    $self->set('StartYear', -4000);
    $self->set('Description', '');
    $self->set('ModPath', '');
}

sub clear {
    my ($self) = @_;
    
    delete $self->{$_} foreach (@fields);

    $self->{'Victories'} = [];
    $self->{'Options'} = [];
    $self->{'MPOptions'} = [];
    $self->{'ForceControls'} = [];
}

sub get {
    my ($self, $key) = @_;
    return $self->{$key};
}

sub set {
    my ($self, $key, $value) = @_;
    $self->{$key} = $value;
}

sub add_victory {
    my ($self, $value) = @_;
    push @{$self->{'Victories'}}, $value;
}

sub remove_victory {
    my ($self, $value) = @_;
    
    my @kept;
    foreach my $victory (@{$self->{'Victories'}}) {
        push @kept, $victory if $victory ne $value;
    }
    
    $self->{'Victories'} = \@kept;
}

sub strip_victories {
    my ($self) = @_;
    
    $self->remove_victory('VICTORY_CONQUEST');
    $self->remove_victory('VICTORY_DOMINATION');
    $self->remove_victory('VICTORY_CULTURAL');
    $self->remove_victory('VICTORY_SPACE_RACE');
    $self->remove_victory('VICTORY_DIPLOMATIC');
}

sub get_victories {
    my ($self, $key) = @_;
    return @{$self->{'Victories'}};
}

sub add_mpoption {
    my ($self, $value) = @_;
    push @{$self->{'MPOptions'}}, $value;
}

sub get_mpoptions {
    my ($self, $key) = @_;
    return @{$self->{'MPOptions'}};
}

sub add_option {
    my ($self, $value) = @_;
    push @{$self->{'Options'}}, $value;
}

sub get_options {
    my ($self, $key) = @_;
    return @{$self->{'Options'}};
}

sub add_force_option {
    my ($self, $value) = @_;
    push @{$self->{'ForceControls'}}, $value;
}

sub get_force_options {
    my ($self, $key) = @_;
    return @{$self->{'ForceControls'}};
}

sub parse {
    my ($self, $fh) = @_;
    
    # begin unit is already found
    while (1) {
        my $line = <$fh>;
        next if $line !~ /\w/;
        next if $line =~ /^\s*#/;
        return if $line =~ /EndGame/i;
        chomp $line;
        $line =~ s/^\s*//;
        $line =~ s/\s*$//;
        
        my ($name, $value) = split '=', $line;
        
        if ($name eq 'Victory') {
            $self->add_victory($value);
        }
        elsif ($name eq 'Option') {
            $self->add_option($value);
        }
        elsif ($name eq 'ForceControl') {
            $self->add_force_option($value);
        }
        elsif ($name eq 'MPOption') {
            $self->add_mpoption($value);
        }
        else {
            $self->set($name, $value);
        }
    }
}

sub write {
    my ($self, $fh) = @_;
    print $fh "BeginGame\n";
    
    foreach my $field (@fields) {
        write_block_data($self, $fh, 1, $field);
        
        if ($field eq 'Calendar') {
            my @options = $self->get_options();
            foreach my $o (@options) {
                $self->set('Option', $o);
                write_block_data($self, $fh, 1, 'Option');
            }
            
            my @mpoptions = $self->get_mpoptions();
            foreach my $o (@mpoptions) {
                $self->set('MPOption', $o);
                write_block_data($self, $fh, 1, 'MPOption');
            }
            
            my @force_options = $self->get_force_options();
            foreach my $o (@force_options) {
                $self->set('ForceControl', $o);
                write_block_data($self, $fh, 1, 'ForceControl');
            }
            
            my @victories = $self->get_victories();
            foreach my $v (@victories) {
                $self->set('Victory', $v);
                write_block_data($self, $fh, 1, 'Victory');
            }
        }
    }
    
    print $fh "EndGame\n";
}


1;