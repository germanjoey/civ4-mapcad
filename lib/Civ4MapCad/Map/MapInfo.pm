package Civ4MapCad::Map::MapInfo;

use strict;
use warnings;

our @fields = ('grid width', 'grid height', 'top latitude', 'bottom latitude', 'wrap X', 'wrap Y', 'world size', 'climate', 'sealevel', 'num plots written', 'num signs written', 'Randomize Resources');
our %field_names;
@field_names{@fields} = (1) x @fields;

use Civ4MapCad::Util qw(write_block_data);

sub new {
    my $proto = shift;
    my $class = ref $proto || $proto;
    
    my $obj = bless { 'wrap X' => 0, 'wrap Y' => 0 }, $class;
    return $obj;
}

sub new_default {
    my $proto = shift;
    my $class = ref $proto || $proto;
    my $obj = bless { 'wrap X' => 0, 'wrap Y' => 0 }, $class;
    
    my ($width, $height) = @_;
    $obj->set_default($width, $height);
    return $obj;
}

sub clear {
    my ($self) = @_;
    delete $self->{$_} foreach (@fields);
    $self->{'Civics'} = [];
}

sub set_default {
    my ($self, $width, $height) = @_;
    
    $self->set('grid width', $width);
    $self->set('grid height', $height);
    $self->set('top latitude', 30);
    $self->set('bottom latitude', 30);
    $self->set('wrap X', 1);
    $self->set('wrap Y', 1);
    $self->set('world size', 'WORLDSIZE_STANDARD');
    $self->set('climate', 'CLIMATE_TEMPERATE');
    $self->set('sealevel', 'SEALEVEL_MEDIUM');
    $self->set('num plots written', $width*$height);
    $self->set('num signs written', 0);
    $self->set('Randomize Resources', 'false');
}

sub get {
    my ($self, $key) = @_;
    return unless exists $self->{$key};
    return $self->{$key};
}

sub set {
    my ($self, $key, $value) = @_;
    
    #if ((($key eq 'wrap X') or ($key eq 'wrap Y')) and ($value == 0)) {
    #    delete $self->{$key};
    #}
    
    $self->{$key} = $value;
}

sub parse {
    my ($self, $fh) = @_;
    
    # begin unit is already found
    while (1) {
        my $line = <$fh>;
        next if $line !~ /\w/;
        next if $line =~ /^\s*#/;
        return if $line =~ /EndMap/i;
        chomp $line;
        $line =~ s/^\s*//;
        $line =~ s/\s*$//;
        
        my ($name, $value) = split '=', $line;
        $self->set($name, $value);
    }
}

sub writeout {
    my ($self, $fh) = @_;
    print $fh "BeginMap\n";
    
    if ((exists $self->{'wrap X'}) and ($self->{'wrap X'} == 0)) {
        delete $self->{'wrap X'};
    }
    
    if ((exists $self->{'wrap Y'}) and ($self->{'wrap Y'} == 0)) {
        delete $self->{'wrap Y'};
    }
    
    foreach my $field (@fields) {
        write_block_data($self, $fh, 1, $field);
    }
    
    print $fh "EndMap\n";
}


1;