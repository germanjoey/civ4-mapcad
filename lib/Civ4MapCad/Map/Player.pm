package Civ4MapCad::Map::Player;

use strict;
use warnings;

our @fields = qw(Team LeaderType LeaderName CivDesc CivShortDesc CivAdjective FlagDecal WhiteFlag CivType Color ArtStyle PlayableCiv MinorNationStatus StartingGold StartingX StartingY StateReligion StartingEra RandomStartLocation Handicap);
our %field_names;
@field_names{@fields} = (1) x @fields;

use Civ4MapCad::Util qw(write_block_data);

sub new {
    my $proto = shift;
    my $class = ref $proto || $proto;
    my $obj = bless {}, $class;
    
    $obj->{'Civics'} = [];
    return $obj;
}

sub new_default {
    my $proto = shift;
    my $class = ref $proto || $proto;
    my $obj = bless {}, $class;
    
    my $teamID = shift;
    $obj->set_default($teamID);
    return $obj;
}

sub set_default {
    my ($self, $teamID) = @_;
    $self->clear();
    $self->set('Team', $teamID);
    $self->set('LeaderType', 'NONE');
    $self->set('CivType', 'NONE');
    $self->set('Color', 'NONE');
    $self->set('ArtStyle', 'NONE');
    $self->set('Handicap', $main::state->{'config'}{'difficulty'});
}

sub clear {
    my ($self) = @_;
    delete $self->{$_} foreach (@fields);
    $self->{'Civics'} = [];
}

sub get {
    my ($self, $key) = @_;
    return $self->{$key};
}

sub set {
    my ($self, $key, $value) = @_;
    $self->{$key} = $value;
}

sub get_civics {
    my ($self) = @_;
    return @{$self->{'Civics'}};
}

sub add_civics {
    my ($self, $line) = @_;
    
    my ($civic_option, $civic) = split ' ', $line;
    $civic_option =~ s/,$//;
    my @civic_option_parts = split '=', $civic_option;
    my @civic_parts = split '=', $civic;
    
    push @{$self->{'Civics'}}, [$civic_option_parts[1], $civic_parts[1]];
}

sub parse {
    my ($self, $fh) = @_;
    
    # begin unit is already found
    while (1) {
        my $line = <$fh>;
        next if $line !~ /\w/;
        next if $line =~ /^\s*#/;
        return if $line =~ /EndPlayer/i;
        chomp $line;
        $line =~ s/^\s*//;
        $line =~ s/\s*$//;
        
        if ($line =~ /^CivicOption/i) {
            $self->add_civics($line);
        }
        
        elsif ($line =~ /^StartingX/) {
            my ($StartingX, $StartingY) = split ',', $line;
            my ($stX, $stXV) = split '=', $StartingX;
            my ($stY, $stYV) = split '=', $StartingY;
            $self->set('StartingX', $stXV);
            $self->set('StartingY', $stYV);
        }
        else {
            my ($name, $value) = split '=', $line;
            $self->set($name, $value);
        }
    }
}

sub writeout {
    my ($self, $fh) = @_;
    print $fh "BeginPlayer\n";
    
    foreach my $field (@fields) {
        next if $field eq 'StartingY';
        if ($field eq 'StartingX') {
            write_block_data($self, $fh, 1, 'StartingX', 'StartingY');
        }
        elsif ($field eq 'Handicap') {
            foreach my $civic ($self->get_civics()) {
                $self->set('CivicOption', $civic->[0]);
                $self->set('Civic', $civic->[1]);
                write_block_data($self, $fh, 1, 'CivicOption', 'Civic');
            }
        
            write_block_data($self, $fh, 1, $field);
        }
        else {
            write_block_data($self, $fh, 1, $field);
        }
        
    }
    
    print $fh "EndPlayer\n";
}

sub is_active {
    my ($self) = @_;
    
    return 1 if $self->{'LeaderType'} ne 'NONE';
    return 0;
}

sub set_from_data {
    my ($self, $data) = @_;
    
    my $teamID = $self->get('Team');
    $self->clear();
    
    # we don't expect every single field here, so only do these
    my @expected_fields = ('CivType', 'CivDesc', 'CivShortDesc', 'CivAdjective', 'Color', 'ArtStyle', 'PlayableCiv', 'WhiteFlag',
                           'MinorNationStatus', 'StartingX', 'StartingY', 'StateReligion', 'RandomStartLocation', 'FlagDecal');
    
    foreach my $key (@expected_fields) {
        $self->set($key, $data->{$key});
    }
            
    $self->set('Handicap', $main::state->{'config'}{'difficulty'});
    
    # assign a random leader for that civ
    my $leader_count = 0 + @{ $data->{'_LeaderType'} };
    my $rand_leader = int( $leader_count*rand(1) );
    $self->set('LeaderType', $data->{'_LeaderType'}[$rand_leader][0]);
    $self->set('LeaderName', $data->{'_LeaderType'}[$rand_leader][1]);
    
    # each civic will be formatted properly in the load_xml_data command
    foreach my $civic (@{ $data->{'_Civics'} }) {
        $self->add_civics($civic);
    }
    
    $self->set('Team', $teamID);
}

1;