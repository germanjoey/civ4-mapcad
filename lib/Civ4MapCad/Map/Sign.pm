package Civ4MapCad::Map::Sign;

use strict;
use warnings;

our @fields = qw(plotX plotY playerType caption);
our %field_names;
@field_names{@fields} = (1) x @fields;

use Civ4MapCad::Util qw(write_block_data);

sub new {
    my $proto = shift;
    my $class = ref $proto || $proto;
    my $obj = bless {}, $class;
    
    return $obj;
}

sub new_default {
    my $proto = shift;
    my $class = ref $proto || $proto;
    my $obj = bless {}, $class;
    
    return $obj;
}

sub clear {
    my ($self) = @_;
    delete $self->{$_} foreach (@fields);
}

sub get {
    my ($self, $key) = @_;
    return $self->{$key};
}

sub set {
    my ($self, $key, $value) = @_;
    $self->{$key} = $value;
}

sub parse {
    my ($self, $fh) = @_;
    
    # begin unit is already found
    while (1) {
        my $line = <$fh>;
        next if $line !~ /\w/;
        next if $line =~ /^\s*#/;
        return if $line =~ /EndSign/i;
        chomp $line;
        $line =~ s/^\s*//;
        $line =~ s/\s*$//;
        
        my ($name, @rest) = split '=', $line;
        $self->set($name, join('=', @rest));
    }
}

sub writeout {
    my ($self, $fh) = @_;
    print $fh "BeginSign\n";
    
    foreach my $field (@fields) {
        write_block_data($self, $fh, 1, $field);
    }
    
    print $fh "EndSign\n";
}


1;