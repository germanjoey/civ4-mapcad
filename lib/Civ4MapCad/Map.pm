package Civ4MapCad::Map;

use strict;
use warnings;

use Civ4MapCad::Util qw(write_block_data deepcopy);

our @fields = qw(TeamID RevealMap);
our %field_names;
@field_names{@fields} = (1) x @fields;

use Civ4MapCad::Map::Game;
use Civ4MapCad::Map::Team;
use Civ4MapCad::Map::Player;
use Civ4MapCad::Map::MapInfo;
use Civ4MapCad::Map::Tile;
use Civ4MapCad::Map::Sign;

sub new {
    my $proto = shift;
    my $class = ref $proto || $proto;
    
    my $obj = {
        'Game' => '',
        'Teams' => {},
        'MapInfo' => '',
        'Players' => [],
        'Tiles' => [], # 2d array
        'Signs' => [],
    };
    
    return bless $obj, $class;
}

sub new_default {
    my $proto = shift;
    my $class = ref $proto || $proto;
    my ($width, $height) = @_;
    
    my $obj = {
        'Version' => 'Version=11',
        'Game' => '',
        'Teams' => {},
        'MapInfo' => '',
        'Players' => [],
        'Tiles' => [], # 2d array
        'Signs' => [],
    };
    $obj = bless $obj, $class;
    
    $obj->default($width, $height);
    return $obj;
}

# TODO: set up a unified interface?
sub info {
    my ($self, $field) = @_;
    return $self->{'MapInfo'}->get($field);
}

sub default {
    my ($self, $width, $height) = @_;
    
    $self->{'Game'} = Civ4MapCad::Map::Game->new_default;
    $self->{'MapInfo'} = Civ4MapCad::Map::MapInfo->new_default($width, $height);
    
    # TODO: num teams is dependent on mod
    foreach my $i (0..17) {
        my $team = Civ4MapCad::Map::Team->new_default($i);
        $self->{'Teams'}{$i} = $team;
        
        my $player = Civ4MapCad::Map::Player->new_default($i);
        push @{$self->{'Players'}}, $player;
    }
    
    foreach my $x (0..$width-1) {
        foreach my $y (0..$height-1) {
            $self->{'Tiles'}[$x][$y] = Civ4MapCad::Map::Tile->new_default($x, $y);
        }
    }
}

sub clear {
    my ($self) = @_;
    
    $self->{'Signs'} = [];
    $self->{'Game'}->clear();
    $self->{'MapInfo'}->clear();
    
    foreach my $team ($self->get_teams()) {
        $team->clear();
        delete $self->{'Teams'}{$team};
    }
    
    foreach my $player ($self->get_players()) {
        $player->clear();
    }
    $self->{'Players'} = [];
    
    $self->clear_map();
}

sub clear_map {
    my ($self) = @_;

    foreach my $x (0..$#{$self->{'Tiles'}}) {
        foreach my $y (0..$#{$self->{'Tiles'}[$x]}) {
            $self->{'Tiles'}[$x][$y]->clear();
            $self->{'Tiles'}[$x][$y]->default($x, $y);
        }
    }
}

sub expand_dim {
    my ($self, $width, $height) = @_;
    
    $self->{'MapInfo'}->set('grid width', $width);
    $self->{'MapInfo'}->set('grid height', $height);
    $self->{'MapInfo'}->set('num plots written', $width*$height);
    # $self->{'MapInfo'}->set('num signs written', $num_signs);
    
    foreach my $x (0..$width-1) {
        $self->{'Tiles'}[$x] = [] unless defined $self->{'Tiles'}[$x];
        foreach my $y (0..$height-1) {
            $self->{'Tiles'}[$x][$y] = Civ4MapCad::Map::Tile->new_default() unless defined $self->{'Tiles'}[$x][$y];
        }
    }
}

sub overwrite_tiles {
    my ($self, $map, $offsetX, $offsetY) = @_;
    
    my $width = $map->info('grid width');
    my $height = $map->info('grid height');
    
    for my $x (0..$width-1) {
        for my $y (0..$height-1) {
            if (! $map->{'Tiles'}[$x][$y]->is_blank()) {
                $self->{'Tiles'}[$x + $offsetX][$y + $offsetY] = deepcopy($map->{'Tiles'}[$x][$y]);
                $self->{'Tiles'}[$x + $offsetX][$y + $offsetY]->set('x', $x + $offsetX);
                $self->{'Tiles'}[$x + $offsetX][$y + $offsetY]->set('y', $y + $offsetX);
            }
        }
    }
}

sub add_player {
    my ($self, $fh) = @_;
    
    my $player = Civ4MapCad::Map::Player->new;
    $player->parse($fh);
    push @{ $self->{'Players'} }, $player;
}

sub get_players {
    my ($self, $fh) = @_;
    return @{$self->{'Players'}};
}

sub add_team {
    my ($self, $fh) = @_;
    
    my $team = Civ4MapCad::Map::Team->new;
    $team->parse($fh);
    my $id = $team->get('TeamID');
    warn "* WARNING: Team $id already exists!" if exists $self->{'Teams'}{$id};
    
    $self->{'Teams'}{$id} = $team;
}

sub get_teams {
    my ($self, $fh) = @_;
    return sort { $a->get('TeamID') <=> $b->get('TeamID') } (values %{$self->{'Teams'}});
}

sub add_sign {
    my ($self, $fh) = @_;
    
    my $sign = Civ4MapCad::Map::Sign->new;
    $sign->parse($fh);
    push @{$self->{'Signs'}}, $sign;
}

sub add_tile {
    my ($self, $fh) = @_;
    
    my $tile = Civ4MapCad::Map::Tile->new;
    $tile->parse($fh);
    
    my $x = $tile->get('x');
    my $y = $tile->get('y');
    
    $self->{'Tiles'}[$x] = [] unless defined $self->{'Tiles'}[$x];
    $self->{'Tiles'}[$x][$y] = $tile;
}

sub fill_tile {
    my ($self, $x, $y) = @_;
    $self->{'Tiles'}[$x][$y]->fill();
}

sub delete_tile {
    my ($self, $x, $y) = @_;
    $self->{'Tiles'}[$x][$y] = Civ4MapCad::Map::Tile->new_default($x, $y);
}

sub set_game {
    my ($self, $fh) = @_;
    $self->{'Game'} = Civ4MapCad::Map::Game->new;
    $self->{'Game'}->parse($fh);
}

sub set_map_info {
    my ($self, $fh) = @_;
    $self->{'MapInfo'} = Civ4MapCad::Map::MapInfo->new;
    $self->{'MapInfo'}->parse($fh);
}

sub import_map {
    my ($self, $filename) = @_;
    
    open (my $fh, $filename) or die $!;
    $self->{'Version'} = <$fh>;
    chomp $self->{'Version'};
    
    while (1) {
        my $line = <$fh>;
        last unless defined $line;
        
        next if $line !~ /\w/;
        next if $line =~ /^\s*#/;
        chomp $line;
        $line =~ s/^\s*//;
        $line =~ s/\s*$//;
        
        if ($line =~ /BeginGame/) {
            $self->set_game($fh);
        }
        elsif ($line =~ /BeginTeam/) {
            $self->add_team($fh);
        }
        elsif ($line =~ /BeginPlayer/) {
            $self->add_player($fh);
        }
        elsif ($line =~ /BeginMap/) {
            $self->set_map_info($fh);
        }
        elsif ($line =~ /BeginPlot/) {
            $self->add_tile($fh);
        }
        elsif ($line =~ /BeginSign/) {
            $self->add_sign($fh);
        }
        else {
            die "Unidentified block found when importing: '$line'";
        }
    }
    
    close $fh;
}

sub export_map {
    my ($self, $filename) = @_;
    
    open (my $fh, '>', $filename) or die $!;
    
    print $fh $self->{'Version'}, "\n";
    
    $self->{'Game'}->write($fh);
    
    foreach my $team ($self->get_teams()) {
        $team->write($fh);
    }
    
    foreach my $player ($self->get_players()) {
        $player->write($fh);
    }
    
    $self->{'MapInfo'}->write($fh);
    print $fh "\n### Plot Info ###\n";

    foreach my $xv (@{ $self->{'Tiles'} }) {
        foreach my $tile (@$xv) {
            $tile->write($fh);
        }
    }
    
    if (@{$self->{'Signs'}} > 0) {
        print $fh "\n### Sign Info ###\n";
    }
    foreach my $sign (@{$self->{'Signs'}}) {
        $sign->write($fh);
    }
}