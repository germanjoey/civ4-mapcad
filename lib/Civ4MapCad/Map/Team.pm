package Civ4MapCad::Map::Team;

use strict;
use warnings;

our @fields = qw(TeamID RevealMap);
our %field_names;
@field_names{@fields} = (1) x @fields;

use Civ4MapCad::Util qw(write_block_data);

sub new {
    my $proto = shift;
    my $class = ref $proto || $proto;
    
    my $obj = bless {}, $class;
    $obj->{'Contacts'} = [];
    $obj->{'Techs'} = [];
    $obj->{'AtWars'} = [];
    $obj->{'PermanentWarPeaces'} = [];
    $obj->{'OpenBordersWithTeams'} = [];
    $obj->{'DefensivePactWithTeams'} = [];
    $obj->{'ProjectTypes'} = [];
    
    return $obj;
}

sub new_default {
    my $proto = shift;
    my $class = ref $proto || $proto;
    my $obj = bless {}, $class;
    
    my $teamID = shift;
    $obj->default($teamID);
    return $obj;
}

sub default {
    my ($self, $teamID) = @_;
    $self->clear();
    $self->set('TeamID', $teamID);
    $self->set('ContactWithTeam', $teamID);
    $self->set('RevealMap', 0);
}

sub clear {
    my ($self) = @_;
    
    delete $self->{$_} foreach (@fields);

    $self->{'Contacts'} = [];
    $self->{'Techs'} = [];
    $self->{'AtWars'} = [];
    $self->{'PermanentWarPeaces'} = [];
    $self->{'OpenBordersWithTeams'} = [];
    $self->{'DefensivePactWithTeams'} = [];
    $self->{'ProjectTypes'} = [];
}

sub get {
    my ($self, $key) = @_;
    return $self->{$key};
}

sub set {
    my ($self, $key, $value) = @_;
    $self->{$key} = $value;
}

sub add_contact {
    my ($self, $value) = @_;
    push @{$self->{'Contacts'}}, $value;
}

sub get_contacts {
    my ($self) = @_;
    return @{$self->{'Contacts'}};
}

sub add_tech {
    my ($self, $value) = @_;
    push @{$self->{'Techs'}}, $value;
}

sub get_techs {
    my ($self) = @_;
    return @{$self->{'Techs'}};
}

sub add_war {
    my ($self, $value) = @_;
    push @{$self->{'AtWars'}}, $value;
}

sub get_wars {
    my ($self) = @_;
    return @{$self->{'AtWars'}};
}

sub add_peace {
    my ($self, $value) = @_;
    push @{$self->{'PermanentWarPeaces'}}, $value;
}

sub get_peaces {
    my ($self) = @_;
    return @{$self->{'PermanentWarPeaces'}};
}

sub add_ob {
    my ($self, $value) = @_;
    push @{$self->{'OpenBordersWithTeams'}}, $value;
}

sub get_obs {
    my ($self) = @_;
    return @{$self->{'OpenBordersWithTeams'}};
}

sub add_pact {
    my ($self, $value) = @_;
    push @{$self->{'DefensivePactWithTeams'}}, $value;
}

sub get_pacts {
    my ($self) = @_;
    return @{$self->{'DefensivePactWithTeams'}};
}

sub add_project {
    my ($self, $value) = @_;
    push @{$self->{'ProjectTypes'}}, $value;
}

sub get_projects {
    my ($self) = @_;
    return @{$self->{'ProjectTypes'}};
}

sub parse {
    my ($self, $fh) = @_;
    
    # begin unit is already found
    while (1) {
        my $line = <$fh>;
        next if $line !~ /\w/;
        next if $line =~ /^\s*#/;
        return if $line =~ /EndTeam/i;
        chomp $line;
        $line =~ s/^\s*//;
        $line =~ s/\s*$//;
        
        my ($name, $value) = split '=', $line;
        
        if ($name eq 'Tech') {
            $self->add_tech($value);
        }
        elsif ($name eq 'ContactWithTeam') {
            $self->add_contact($value);
        }
        elsif ($name eq 'AtWar') {
            $self->add_war($value);
        }
        elsif ($name eq 'PermanentWarPeace') {
            $self->add_peace($value);
        }
        elsif ($name eq 'OpenBordersWithTeam') {
            $self->add_ob($value);
        }
        elsif ($name eq 'DefensivePactWithTeam') {
            $self->add_pact($value);
        }
        elsif ($name eq 'ProjectType') {
            $self->add_project($value);
        }
        else {
            $self->set($name, $value);
        }
    }
}

sub write {
    my ($self, $fh) = @_;
    print $fh "BeginTeam\n";
    
    write_block_data($self, $fh, 1, 'TeamID');
    
    my @techs = $self->get_techs();
    foreach my $t (@techs) {
        $self->set('Tech', $t);
        write_block_data($self, $fh, 1, 'Tech');
    }
    
    my @contacts = $self->get_contacts();
    foreach my $c (@contacts) {
        $self->set('ContactWithTeam', $c);
        write_block_data($self, $fh, 1, 'ContactWithTeam');
    }
    
    my @wars = $self->get_wars();
    foreach my $w (@wars) {
        $self->set('AtWar', $w);
        write_block_data($self, $fh, 1, 'AtWar');
    }
    
    my @peaces = $self->get_peaces();
    foreach my $p (@peaces) {
        $self->set('PermanentWarPeace', $p);
        write_block_data($self, $fh, 1, 'PermanentWarPeace');
    }
    
    my @obs = $self->get_obs();
    foreach my $o (@obs) {
        $self->set('OpenBordersWithTeam', $o);
        write_block_data($self, $fh, 1, 'OpenBordersWithTeam');
    }
    
    my @pacts = $self->get_pacts();
    foreach my $p (@pacts) {
        $self->set('DefensivePactWithTeam', $p);
        write_block_data($self, $fh, 1, 'DefensivePactWithTeam');
    }
    
    my @projects = $self->get_projects();
    foreach my $p (@projects) {
        $self->set('ProjectType', $p);
        write_block_data($self, $fh, 1, 'ProjectType');
    }
    
    write_block_data($self, $fh, 1, 'RevealMap');
        
    print $fh "EndTeam\n";
}


1;