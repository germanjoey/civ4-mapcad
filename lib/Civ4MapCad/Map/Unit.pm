package Civ4MapCad::Map::Unit;

our @fields = qw(UnitType UnitOwner UnitName Damage Level Experience FacingDirection UnitAIType PromotionType);
our %field_names;
@field_names{@fields} = (1) x @fields;

use Civ4MapCad::Util qw(write_block_data);

sub new {
    my $proto = shift;
    my $class = ref $proto || $proto;
    
    my $obj = bless {}, $class;
    $obj->{'Promotions'} = [];
    
    return $obj;
}

sub add_promotion {
    my ($self, $val) = @_;
    push @{$self->{'promotions'}}, $val;
}

sub get_promotions {
    my ($self) = @_;
    return @{ $self->{'promotions'} };
}

sub get {
    my ($self, $key) = @_;
    return $self->{$key};
}

sub set {
    my ($self, $key, $value) = @_;
    $self->{$key} = $value;
}
    
sub parse {
    my ($self, $fh) = @_;
    
    # begin unit is already found
    while (1) {
        my $line = <$fh>;
        next if $line !~ /\w/;
        next if $line =~ /^\s*#/;
        return if $line =~ /EndUnit/i;
        chomp $line;
        $line =~ s/^\s*//;
        $line =~ s/\s*$//;
        
        my @pieces = split ',', $line;
        foreach my $piece (@pieces) {
            $piece =~ s/,$//;
            $piece =~ s/^\s*//;
            $piece =~ s/\s*$//;
            
            my ($name, $value) = split '=', $piece;
            
            if ($name eq 'PromotionType') {
                $self->add_promotion($value);
            }
            else {
                $self->set($name, $value);
            }
        }
    }
}

sub write {
    my ($self, $fh) = @_;
    print $fh "\tBeginUnit\n";
    
    write_block_data($self, $fh, 2, 'UnitType', 'UnitOwner');
    write_block_data($self, $fh, 2, 'UnitName');
    write_block_data($self, $fh, 2, 'Damage');
    write_block_data($self, $fh, 2, 'Level', 'Experience');
    
    my @promotions = $self->get_promotions();
    foreach my $promo (@promotions) {
        $self->set('PromotionType', $promo);
        write_block_data($self, $fh, 2, 'PromotionType');
    }
    
    write_block_data($self, $fh, 2, 'FacingDirection');
    write_block_data($self, $fh, 2, 'UnitAIType');
    
    print $fh "\tEndUnit\n";
}

1;