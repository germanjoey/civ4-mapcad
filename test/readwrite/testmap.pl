#!perl

use strict;
use warnings;

use lib '../../lib';

use Civ4MapCad::Map::MapInfo;
my $map = Civ4MapCad::Map::MapInfo->new();

my $filename = 'testMap.CivBeyondSwordWBSave';

open (my $fhi, $filename) or die $!;
my $beginMap = <$fhi>;
$map->parse($fhi);
close $fhi;

use Data::Dumper;
print Dumper $map;

open (my $fho, '>', "$filename.out") or die $!;
$map->write($fho);
close $fho;