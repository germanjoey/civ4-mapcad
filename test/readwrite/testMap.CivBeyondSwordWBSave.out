BeginMap
	grid width=68
	grid height=45
	top latitude=45
	bottom latitude=45
	wrap X=1
	wrap Y=1
	world size=WORLDSIZE_STANDARD
	climate=CLIMATE_TEMPERATE
	sealevel=SEALEVEL_MEDIUM
	num plots written=3060
	num signs written=1
	Randomize Resources=false
EndMap
